2dlib/CMakeLists.txt

```diff
@@ -13,3 +13,12 @@ add_library(2dlib STATIC ${HEADERS} ${CPPS})
 target_link_libraries(2dlib PRIVATE
   cfile
 )
+
+target_include_directories(2dlib PUBLIC
+  $<BUILD_INTERFACE:
+    ${PROJECT_SOURCE_DIR}/bitmap
+  >
+  $<BUILD_INTERFACE:
+    ${PROJECT_SOURCE_DIR}/renderer
+  >
+)
```

Comment:
This should be not target_include_directories, but target_link_libraries (with prevoius one):
```
target_link_libraries(2dlib PRIVATE
  bitmap
  cfile
  renderer
)
```

After that you should declare build interface of module itself

```
target_include_directories(2dlib PUBLIC
  $<BUILD_INTERFACE:
    ${PROJECT_SOURCE_DIR}/2dlib
  >
)
```

Same goes to all other target_include_directories changes. See cfile submodule layout as common example how it should looks like.

Reply:
> After that you should declare build interface of module itself

This is incorrect because it's translated to `-Idirectory` which is a search path.

Reply:
> This should be not target_include_directories, but target_link_libraries (with prevoius one):

Doing so would result in overlapping/recursive linking dependencies.

Reply:
> This should be not target_include_directories, but target_link_libraries (with prevoius one):

This creates an impossible situation because you are recursively linking. This infinite recursions ends on a seemingly arbitrary subproject. In this case, it tried to compile `AudioEncode` when I requested `linux` to be built.

``` sh
FAILED: AudioEncode/CMakeFiles/AudioEncode.dir/Debug/audio_encode.cpp.o 
/usr/bin/c++ -DCHECKSUM=2273889835UL -DLOGGER -DPRIMARY_HOG=\"d3-linux.hog\" -D_DEBUG -D_USE_OGL_ACTIVE_TEXTURES -D__LINUX__ -DCMAKE_INTDIR=\"Debug\" -I/home/gravis/project/Descent3/cfile -I/home/gravis/project/Descent3/ddebug -I/home/gravis/project/Descent3/fix -I/home/gravis/project/Descent3/lib -I/home/gravis/project/Descent3/linux -I/home/gravis/project/Descent3/Descent3 -I/home/gravis/project/Descent3/lib/linux -I/usr/include/SDL2 -g -fPIC -Wno-multichar -std=c++17 -MD -MT AudioEncode/CMakeFiles/AudioEncode.dir/Debug/audio_encode.cpp.o -MF AudioEncode/CMakeFiles/AudioEncode.dir/Debug/audio_encode.cpp.o.d -o AudioEncode/CMakeFiles/AudioEncode.dir/Debug/audio_encode.cpp.o -c /home/gravis/project/Descent3/AudioEncode/audio_encode.cpp
/home/gravis/project/Descent3/AudioEncode/audio_encode.cpp:25:10: fatal error: aencode.h: No such file or directory
   25 | #include "aencode.h"                                                                                                                                          
      |          ^~~~~~~~~~~                                                                                                                                                                  
compilation terminated.                   
```
`AudioEncode` needs `aencode.h` and here's the `CMakeLists.txt`

``` sh
set(HEADERS
  audio_encode.h)
set(CPPS
  audio_encode.cpp)

add_library(AudioEncode STATIC ${HEADERS} ${CPPS})
target_link_libraries(AudioEncode PUBLIC
  libacm
)
```
Notice that there is no `-I` for libacm in the first `g++` command despite being a linking requirement.  Just in case you needed to know, here' the directory:

``` sh
~/project/Descent3$ ls libacm/
adecode.cpp  adecode.h  aencode.cpp  aencode.h  CMakeLists.txt  decode.c  libacm.h
```

**I cannot do what you ask because it will not compile.**

Reply:
You didn't `target_include_directory` on submodule. Again **see cfile for the reference**.

```
set(CPPS
  encoder.cpp
)

add_library(AudioEncode STATIC ${CPPS})
target_link_libraries(AudioEncode libacm)
target_include_directories(AudioEncode PUBLIC
  $<BUILD_INTERFACE:
  ${PROJECT_SOURCE_DIR}/AudioEncode
  >
)
```

```

  Checking Build System
  Building Custom Rule C:/Users/azama/CLionProjects/Descent3/libacm/CMakeLists.txt
  aencode.cpp
  adecode.cpp
  decode.c
  libacm.vcxproj -> C:\Users\azama\CLionProjects\Descent3\builds\win32\libacm\Debug\libacm.lib
  Building Custom Rule C:/Users/azama/CLionProjects/Descent3/AudioEncode/CMakeLists.txt
  encoder.cpp
  AudioEncode.vcxproj -> C:\Users\azama\CLionProjects\Descent3\builds\win32\AudioEncode\Debug\AudioEncode.lib

Build finished
```

Reply:
I can't believe it worked.